port/cpl_vsil_cache.cpp

```diff
@@ -360,8 +360,15 @@ size_t VSICachedFile::Read(void *pBuffer, size_t nSize, size_t nCount)
     /*      Make sure the cache is loaded for the whole request region.     */
     /* ==================================================================== */
     const vsi_l_offset nStartBlock = m_nOffset / m_nChunkSize;
-    const vsi_l_offset nEndBlock =
-        (m_nOffset + nRequestedBytes - 1) / m_nChunkSize;
+    // Calculate last block
+    const vsi_l_offset nLastBlock = m_nFileSize / m_nChunkSize;
+    vsi_l_offset nEndBlock = (m_nOffset + nRequestedBytes - 1) / m_nChunkSize;
+
+    // if nLastBlock is not 0 consider the min value to avoid out-of-range reads
+    if (nLastBlock != 0 && nEndBlock > nLastBlock)
+    {
+        nEndBlock = nLastBlock;
```

Comment:
is this code path exercised by existing tests?

Reply:
I don't know. I was thinking at a test but I have no idea where to reproduce this scenario with vsicache and vsicurl in a test.

Reply:
Perhaps add an assertion and run autotest/gcore/vsifile.py ? There are tests in it with VSI_CACHE=YES

> but I have no idea how to reproduce this scenario with vsicache and vsicurl in a test.

Testing on a regular file should be enough

Reply:
ok, done. But without using curl there is no way to catch the error message that was triggered by curl, so the test will always pass. 
The new code path is triggered by the test though.

Reply:
I was thinking to a simpler test that would not involve zip or tiff, but just use the VSILFILE API and issue a gdal.VSIFReadL() with a size bigger than the the file size.

Something like:
```
f = gdal.VSIFOpenL(tmp_filename, "wb")
gdal.VSIFWriteL(b"x" * 100000, 100000, 1, f)
gdal.VSIFCloseL(f)
with gdal.config_option("VSI_CACHE", "YES"):
    f = gdal.VSIFReadL(tmp_filename, "rb")
    gdal.VSIFSeekL(f, 60000, 0)
    data = gdal.VSIFReadL(1, 75000, f)  # reads past end of file
    gdal.VSIFCloseL(f)
    assert data == b"x" * 40000
```

Reply:
ok, test updated and verified that it triggers the new code path.